### PR TITLE
npeg based parser

### PR DESCRIPTION
**Pull request by Menduist**
The current parser has many flaws which makes it incompatible with the performance suite.
Even the author wants to switch it to npeg: https://forum.nim-lang.org/t/9036#58930

so here it is!
Todo:
- [X] Lexer
- [X] Parser
- [x] Import handling
- [x] Check conformity better
- [x] Cleanup
- [x] Protobuf2?
- [x] https://github.com/zevv/npeg/pull/65

Fun fact: the [official protobuf syntax](https://developers.google.com/protocol-buffers/docs/reference/proto3-spec) is invalid and out of date! :tada: (doesn't allow leading underscore for field names, doesn't handle `optional`, etc)

### CODE REVIEW

**Comment by arnetheduck** [comment]
> invalid and out of date!

keep in mind that `optional` is a pb2 feature, it doesn't exist in pb3 that you linked - perhaps that's why it's.. not working that well ;)

that said, any parser will have a negative impact on compile times - specially in "core libraries" that get recompiled over and over, it probably will remain a bad idea to use the `.proto` files.

**Comment by Menduist** [comment]
`optional` was "back-ported" to protobuf3 AFAICT: https://developers.google.com/protocol-buffers/docs/proto3#specifying_field_rules

The `.proto` -> structure macro is useful for a few reason:
- For conformance & stress testing
- To make sure that your structure is valid & up to date.

In my ideal world, we would have something like:
```nim
import_proto3 "somefile.proto":
  Cookie {.proto3.} = object
    offset {.fieldNumber: 1, pint.}: uint64
    ns {.fieldNumber: 2.}: string
  [..]
```
In regular compilation, `import_proto3` would be a noop.
However, in CI, we would be able to check that our definition is correct & up to date (with easy copy/paste to fix if that's not the case)

Seems like this is the best of all worlds (don't have to guess what the structure is, no macro generating hided stuff, but still up to date etc). Still need to think about it though

**Comment by arnetheduck** [comment]
right you are, looks like I'm out of date - they even wrote a doc about it: https://github.com/protocolbuffers/protobuf/blob/main/docs/implementing_proto3_presence.md

File: protobuf_serialization/files/proto_parser.nim [hunk 1]
```diff
@@ -1,232 +1,240 @@
...
+
+proc `==`(t: Token, c: char): bool =
+  t.typ == Symbol and t.text[0] == c
+
+proc tokenize(f: string): seq[Token] =
+  let lexer = peg(tokens, res: seq[Token]):
+    space    <- +Space
+    comment  <- "//" * *(1 - '\n')
+    comment2 <- "/*" * @"*/"
+    dblstring<- '"' * *(1 - '"') * '"'
```
Comment (lchenut) [inline protobuf_serialization/files/proto_parser.nim]: If I understand correctly, `"\""` will not be correctly tokenized.

Comment (Menduist) [reply protobuf_serialization/files/proto_parser.nim]: Yup, need to implement this https://developers.google.com/protocol-buffers/docs/reference/proto3-spec#string_literals

**Comment by arnetheduck** [comment]
Tracking in https://github.com/status-im/nim-protobuf-serialization/issues/42

**Review by lchenut — APPROVED** [review]

**Pull request merged**